src/main/java/org/openrewrite/maven/AbstractRewriteMojo.java

```diff
@@ -65,12 +65,16 @@ public abstract class AbstractRewriteMojo extends ConfigurableRewriteMojo {
     private static final String RECIPE_NOT_FOUND_EXCEPTION_MSG = "Could not find recipe '%s' among available recipes";
 
     protected Environment environment() throws MojoExecutionException {
+        return environment(getRecipeArtifactCoordinatesClassloader());
+    }
+
+    protected Environment environment(@Nullable ClassLoader recipeClassLoader) throws MojoExecutionException {
         Environment.Builder env = Environment.builder(project.getProperties());
-        if (getRecipeArtifactCoordinates().isEmpty()) {
+        if (recipeClassLoader == null) {
             env.scanRuntimeClasspath()
                     .scanUserHome();
         } else {
-            env.load(new ClasspathScanningLoader(project.getProperties(), getRecipeClassloader()));
+            env.load(new ClasspathScanningLoader(project.getProperties(), recipeClassLoader));
         }

```

Comment:
With these changes, we should get the same result just running `env.scanRuntimeClasspath` unconditionally here, but I kept this `if/else` for now to ensure the existing behavior is preserved if we run into the ClassCastException risked in the `merge` method